include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,414 @@
+    {
+        this->cellDescription = cellDescription;
+    }
+
+    void restart(uint32_t timeStep, const std::string restartDirectory)
+    {
+        // Keep this empty
+    }
+
+    void checkpoint(uint32_t timeStep, const std::string restartDirectory)
+    {
+        // Keep this empty
+    }
+
+  private:
+    /**
+     * The plugin is loaded on every MPI rank, and therefore this function is
+     * executed on every MPI rank.
+     * One host with MPI rank 0 is defined to be the master.
+     * It creates a folder where all the
+     * results are saved in a plain text format.
+     **/
+    void pluginLoad()
+    {
+        if (!notifyPeriod.empty())
+        {
+            isMaster = reduce.hasResult(mpi::reduceMethods::Reduce());
+            n_q = n_qx * n_qy * n_qz;
+            sumfcoskr = new GridBuffer<float1_64, DIM1>(DataSpace<DIM1>(n_q));
+            sumfsinkr = new GridBuffer<float1_64, DIM1>(DataSpace<DIM1>(n_q));
+            // allocate one float on GPU and host
+            np = new GridBuffer<float1_64, DIM1>(DataSpace<DIM1>(1));
+            // allocate one int on GPU and host
+            nmp = new GridBuffer<int64_t, DIM1>(DataSpace<DIM1>(1));
+            sumfcoskr_master = new float1_64[n_q];
+            sumfsinkr_master = new float1_64[n_q];
+            intensity_master = new float1_64[n_q];
+            Environment<>::get().PluginConnector().setNotificationPeriod(
+                this, notifyPeriod);
+            pmacc::Filesystem<simDim> &fs =
+                Environment<simDim>::get().Filesystem();
+
+            // only rank 0 create a file
+            if (isMaster)
+            {
+                fs.createDirectory("saxsOutput");
+                fs.setDirectoryPermissions("saxsOutput");
+            }
+        }
+    }
+
+    /**
+     * Write scattering intensity for each q
+     * @param intensity
+     * @param name The name of output file
+     **/
+    void writeIntensity(float1_64 *intensity, std::string name)
+    {
+        std::ofstream ofile;
+        ofile.open(name.c_str(), std::ofstream::out | std::ostream::trunc);
+        if (!ofile)
+        {
+            std::cerr << "Can't open file [" << name
+                      << "] for output, disable plugin output.\n";
+            isMaster =
+                false; // no Master anymore -> no process is able to write
+        }
+        else
+        {
+            int i_x, i_y, i_z;
+            float3_64 q;
+
+            ofile << n_q << "\n";
+            ofile << "# qx qy qz intensity \n";
+            for (unsigned int i = 0; i < n_q; i++)
+            {
+                i_z = i % n_qz;
+                i_y = (i / n_qz) % n_qy;
+                i_x = i / (n_qz * n_qy);
+                q[0] = q_min[0] + q_step[0] * i_x;
+                q[1] = q_min[1] + q_step[1] * i_y;
+                q[2] = q_min[2] + q_step[2] * i_z;
+                ofile << q[0] << " " << q[1] << " " << q[2] << " "
```

Comment:
There is an overloaded `operator <<` for vectors. (Not sure if you did not use it intentionally).